### DISCUSSION TITLE
Sending large byte arrays efficiently

### DISCUSSION THREAD

**Discussion by klasyc** [nlohmann/json#3952] [Q&A]
Hello guys,

I have a C++ application with JSON web API ([cpp-httplib](https://github.com/yhirose/cpp-httplib) and [nlohmann/json](https://github.com/nlohmann/json) libraries) where I need to send each second approximately 8 KB long array of data. My idea is to encode the array of bytes with base64 encoding and then put it to the JSON data as a normal string.

Since the application is running on a resource constrained embedded computer, I am thinking about efficiency - mainly, I would like to prevent the application from useless buffer copying. I have two questions:

1. The cpp-httplib offers `DataSink` object, which has a very similar API to the [output adapters](https://github.com/nlohmann/json/blob/master/include/nlohmann/detail/output/output_adapters.hpp) used in the JSON library. If I understand it correctly, it should be possible to create my own output adapter and then serialize JSON data directly to the `DataSink` instead of using a temporary string variable. I tried to implement my own adapter by inheriting the `output_adapter_protocol` struct, but I cannot figure out how to use it. Are there any examples?

2. I would like to create a custom serializer for my array of bytes. Of course, I could convert the array to string first and then use the standard serializer, but I think the most elegant solution would be to make the conversion directly. I found some information about [custom serializers](https://json.nlohmann.me/features/arbitrary_types/#how-can-i-use-get-for-non-default-constructiblenon-copyable-types), but it seems they all convert custom data types to JSON string objects, which adds the unwanted extra step during the conversion. Any ideas?

Thank you in Advance!
Klasyc

**Comment by nlohmann** [comment]
1) The public API (json::dump(), to_msgpack(), to_cbor(), etc.) only accepts detail::output_adapter<CharType>, whose constructors are limited to std::vector, std::basic_ostream, or std::basic_string (see include/nlohmann/detail/output/output_adapters.hpp) — there's no public constructor that takes an arbitrary output_adapter_protocol subclass like cpp-httplib's DataSink directly, since output_adapter/output_adapter_protocol live in the detail:: namespace and aren't meant as a public extension point. The practical route is to wrap DataSink in a custom std::streambuf (forwarding overflow()/xsputn() calls to DataSink::write) and construct a std::ostream from it, then pass that ostream to j.dump() via the `<<` operator or the ostream-based to_msgpack/to_cbor overload — that avoids an extra intermediate std::string without touching library internals. 2) For a custom byte-array serializer that skips the string round-trip, write an adl_serializer<YourByteArrayType>::to_json that assigns directly into a json::binary_t (e.g. j = json::binary(bytes)) rather than converting to a string first — binary_t is written natively (as raw bytes) by the binary formats (CBOR/MessagePack/UBJSON/BJData), so no extra encoding step is needed as long as you're using one of those formats, not plain JSON text (which would still need base64 or similar).

---
*This reply was drafted by [Claude Code](https://claude.com/claude-code) on behalf of @nlohmann, as part of a review of open discussions.*
